modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+            bufferOffset += block.size;
+            block = block.next;
+          }
+
+          // Yield this chunk to be process...
+          segment.releaseStatus();
+          yield (bufferSz, buffer);
+          c_free(buffer);
+        }
+      }
+    }
+  }
+
+  iter these(param tag : iterKind, followThis) where tag == iterKind.follower {
+    var (bufferSz, buffer) = followThis;
+    for i in 0 .. #bufferSz {
+      yield buffer[i];
+    }
+  }
+}
+
+/*
+  A segment block is an unrolled linked list node that holds a contiguous buffer
+  of memory. Each segment block size *should* be a power of two, as we increase the
+  size of each subsequent unroll block by twice the size. This is so that stealing
+  work is faster in that majority of elements are confined to one area.
+*/
+pragma "no doc"
+class BagSegmentBlock {
+  type eltType;
+
+  // Contiguous memory containing all elements
+  var elems :  c_ptr(eltType);
+  var next : BagSegmentBlock(eltType);
+
+  // The capacity of this block.
+  var cap : int;
+  // The number of occupied elements in this block.
+  var size : int;
+
+  inline proc isEmpty {
+    return size == 0;
+  }
+
+  inline proc isFull {
+    return size == cap;
+  }
+
+  inline proc push(elt : eltType) {
+    if elems == nil {
+      halt("DistributedBag Internal Error: 'elems' is nil");
+    }
+
+    if isFull {
+      halt("DistributedBag Internal Error: SegmentBlock is Full");
+    }
+
+    elems[size] = elt;
+    size = size + 1;
+  }
+
+  inline proc pop() : eltType {
+    if elems == nil {
+      halt("DistributedBag Internal Error: 'elems' is nil");
+    }
+
+    if isEmpty {
+      halt("DistributedBag Internal Error: SegmentBlock is Empty");
+    }
+
+    size = size - 1;
+    var elt = elems[size];
+    return elt;
+  }
+
+  proc BagSegmentBlock(type eltType, capacity) {
+    if capacity == 0 {
+      halt("DistributedBag Internal Error: Capacity is 0...");
+    }
+
+    cap = capacity;
+    elems = c_malloc(eltType, capacity);
+  }
+
+  proc BagSegmentBlock(type eltType, ptr, capacity) {
+    cap = capacity;
+    elems = ptr;
+    size = cap;
+  }
+
+  proc ~BagSegmentBlock() {
+    c_free(elems);
+  }
+}
+
+/*
+  A segment is, in and of itself an unrolled linked list. We maintain one per core
+  to ensure maximum parallelism.
+*/
+pragma "no doc"
+record BagSegment {
+  type eltType;
+
+  // Used as a test-and-test-and-set spinlock.
+  var status : atomic uint;
+
+  var headBlock : BagSegmentBlock(eltType);
+  var tailBlock : BagSegmentBlock(eltType);
+
+  var nElems : atomic uint;
+
+  inline proc isEmpty {
+    return nElems.read() == 0;
+  }
+
+  inline proc acquireWithStatus(newStatus) {
+    return status.compareExchangeStrong(STATUS_UNLOCKED, newStatus);
+  }
+
+  // Set status with a test-and-test-and-set loop...
+  inline proc acquire(newStatus) {
+    while true {
+      if currentStatus == STATUS_UNLOCKED && acquireWithStatus(newStatus) {

```

Comment:
I can convert `acquireWithStatus` to use `compareExchangeWeak`, sure, or is this about using the 'Test-And-Test-And-Set' over the 'Test-And-Set' pattern?